src/Shared/Recruit.Vacancies.Client/Infrastructure/Reports/ReportService.cs

```diff
@@ -0,0 +1,104 @@
+            _reportRepository = reportRepository;
+            _reportStrategyAccessor = reportStrategyAccessor;
+            _timeProvider = timeProvider;
+            _csvBuilder = csvBuilder;
+        }
+
+        public async Task GenerateReportAsync(Guid reportId)
+        {
+            var report = await _reportRepository.GetReportAsync(reportId);
+
+            if (report == null)
+            {
+                _logger.LogInformation("Report: {reportId} not found. Ignoring.", reportId);
+                return;
+            }
+
+            if (report.Status == ReportStatus.Generated)
+            {
+                _logger.LogInformation("Report: {reportId} already generated. Ignoring.", reportId);
+                return;
+            }
+
+            try
+            {
+                _logger.LogInformation("Report: {reportId} processing.", reportId);
+                report.Status = ReportStatus.InProgress;
+                report.GenerationStartedOn = _timeProvider.Now;
+
+                await _reportRepository.UpdateAsync(report);
+                _logger.LogInformation("Report: {reportId} in progress.", reportId);
+
+                var reportStrategy = _reportStrategyAccessor(report.ReportType);
+                var data = await reportStrategy.GetReportDataAsync(report.Parameters);
+
+                report = await _reportRepository.GetReportAsync(reportId);
+                if (report.Status == ReportStatus.Generated)
+                {
+                    _logger.LogInformation("Report: {reportId} already generated in the meantime. Ignoring.", reportId);
+                    return;
+                }
+                report.Data = data;
+                report.Status = ReportStatus.Generated;
+                report.GeneratedOn = _timeProvider.Now;
+
+                await _reportRepository.UpdateAsync(report);
+                _logger.LogInformation("Report: {reportId} generated.", reportId);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogWarning(ex, "Failed to generate report: {reportId}", report.Id);
+                report.Status = ReportStatus.Failed;
+                await _reportRepository.UpdateAsync(report);
+                throw;
+            }
+        }
+
+        public void WriteReportAsCsv(Stream stream, Report report)
+        {
+            if (report.Status != ReportStatus.Generated)
+            {
+                throw new Exception($"Cannot download report: {report.Id} as incorrect status: {report.Status}");
+            }
+
+            var reportStrategy = _reportStrategyAccessor(report.ReportType);
+
+            var results = JArray.Parse(report.Data);
+
+            _csvBuilder.WriteCsvToStream(stream, results, report.RequestedOn, reportStrategy.ResolveFormat);
+        }
+
+        public Task IncrementReportDownloadCountAsync(Guid reportId)
+        {
+            return _reportRepository.IncrementReportDownloadCountAsync(reportId);
```

Comment:
This is a bit of a sinkhole, could we not call the repo for the increment method right from the Client

Reply:
👍